experimental/parser/lexer.go

```diff
@@ -0,0 +1,675 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/bufbuild/protocompile/experimental/report"
+	"github.com/bufbuild/protocompile/experimental/token"
+	"github.com/rivo/uniseg"
+)
+
+// Lex performs lexical analysis on the file contained in ctx, and appends any
+// diagnostics that results in to l.
+func Lex(ctx token.Context, errs *report.Report) {
+	lexer := &lexer{
+		Context: ctx,
+		Stream:  ctx.Stream(),
+		Report:  errs,
+	}
+	lexer.Lex()
+}
+
+// lexer is a Protobuf lexer.
+type lexer struct {
+	token.Context
+	*token.Stream // Embedded so we don't have to call Stream() everywhere.
+	*report.Report
+
+	// This is outlined so that it's easy to print in the panic handler.
+	lexerState
+}
+
+type lexerState struct {
+	cursor, count int
+	openStack     []token.Token
+}
+
+// Lex performs lexical analysis, and places any diagnostics in report.
+func (l *lexer) Lex() {
+	defer func() {
+		if panicked := recover(); panicked != nil {
+			panic(fmt.Sprintf("panic while lexing: %s; %#v", panicked, l.lexerState))
+		}
+	}()
+
+	// Check that the file isn't too big. We give up immediately if that's
+	// the case.
+	if len(l.Text()) > MaxFileSize {
+		l.Error(ErrFileTooBig{Path: l.Path()})
+		return
+	}
+
+	// Also check that the text of the file is actually UTF-8.
+	// We go rune by rune to find the first invalid offset.
+	for text := l.Text(); text != ""; {
+		r := decodeRune(text)
+		if r == -1 {
+			l.Error(ErrNotUTF8{
+				Path: l.Path(),
+				At:   len(l.Text()) - len(text),
+				Byte: text[0],
+			})
+			return
+		}
+		text = text[utf8.RuneLen(r):]
```

Comment:
Will `utf8.RuneLen` _always_ be correct? Just want to make sure whether it's possible for an encoder to marshal a rune in a "non-canonical" way, where we'd really need the original length from `utf8.DecodeRune`.

Reply:
UTF-8 does not permit over-long encodings, which DecodeRune rejects.